launch/launch/utilities/signal_management.py

```diff
@@ -12,133 +12,157 @@
-            elif callable(__original_sigquit_handler):
-                __original_sigquit_handler(signum, frame)
-
-    def __on_sigterm(signum, frame):
-        if callable(__custom_sigterm_handler):
-            __custom_sigterm_handler(signum, frame, __original_sigterm_handler)
-        elif callable(__original_sigterm_handler):
-            __original_sigterm_handler(signum, frame)
-
-    # signals must be registered in the main thread, but print a nicer message if we're not there
-    try:
-        signal.signal(signal.SIGINT, __on_sigint)
-        signal.signal(signal.SIGTERM, __on_sigterm)
-        if platform.system() != 'Windows':
-            # Windows does not support SIGQUIT
-            signal.signal(signal.SIGQUIT, __on_sigquit)
-    except ValueError:
-        logger = launch.logging.get_logger(__name__)
-        logger.error("failed to set signal handlers in '{}'".format(__name__))
-        logger.error('this function must be called in the main thread')
-        raise
+    def __init__(
+        self,
+        loop: asyncio.AbstractEventLoop
+    ):
+        """
+        Instantiate manager.
+
+        :param loop: event loop that will handle the signals.
+        """
+        self.__loop = loop  # type: asyncio.AbstractEventLoop
+        self.__background_loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.__handlers = {}  # type: dict
+        self.__prev_wakeup_handle = -1  # type: Union[int, socket.socket]
+        self.__wsock, self.__rsock = socket.socketpair()  # type: Tuple[socket.socket, socket.socket]  # noqa
+        self.__wsock.setblocking(False)
+        self.__rsock.setblocking(False)
+
+    def __enter__(self):
+        try:
+            self.__loop.add_reader(self.__rsock.fileno(), self.__handle_signal)
+        except NotImplementedError:
+            # Some event loops, like the asyncio.ProactorEventLoop
+            # on Windows, do not support asynchronous socket reads.
+            # So we emulate it.
+            self.__background_loop = asyncio.SelectorEventLoop()
+            self.__background_loop.add_reader(
+                self.__rsock.fileno(),
+                self.__loop.call_soon_threadsafe,
+                self.__handle_signal)
+
+            def run_background_loop():
+                asyncio.set_event_loop(self.__background_loop)
+                self.__background_loop.run_forever()
+
+            self.__background_thread = threading.Thread(target=run_background_loop)
+            self.__background_thread.start()
+        self.__prev_wakeup_handle = signal.set_wakeup_fd(self.__wsock.fileno())
+        if self.__prev_wakeup_handle != -1 and is_winsock_handle(self.__prev_wakeup_handle):
+            # On Windows, os.write will fail on a WinSock handle. There is no WinSock API
+            # in the standard library either. Thus we wrap it in a socket.socket instance.
+            self.__prev_wakeup_handle = socket.socket(fileno=self.__prev_wakeup_handle)
```

Comment:
maybe always wrapping the previous `fd` in a `socket.socket` works, without testing `is_winsock_handle`

Reply:
Thing is, wakeup file descriptors might be regular files too.

Reply:
makes sense